File: ticket_booking.js
/**
 * Experiment-3: Concurrent Ticket Booking System with Seat Locking using Redis
 *
 * Stack : Node.js + ioredis
 * Install: npm install ioredis
 * Run   : node ticket_booking.js
 *
 * Concepts demonstrated:
 *  - Atomic seat locking with SET NX PX (set-if-not-exists + TTL)
 *  - Lua scripts for atomic check-and-confirm
 *  - Concurrent booking simulation with Promise.all
 *  - Auto-release of expired locks
 *  - Booking confirmation stored as Redis Hash
 */

const Redis = require('ioredis');
const redis = new Redis(); // connects to localhost:6379 by default

// ─── CONFIG ──────────────────────────────────────────────────────────────────
const LOCK_TTL_MS  = 10 * 1000;  // seat held for 10 seconds
const EVENT_ID     = 'EVT-CONCERT-2026';
const TOTAL_SEATS  = 10;

// ─── KEY HELPERS ─────────────────────────────────────────────────────────────
const seatLockKey    = (seat) => `lock:${EVENT_ID}:seat:${seat}`;
const seatStatusKey  = (seat) => `status:${EVENT_ID}:seat:${seat}`;
const bookingKey     = (bookingId) => `booking:${bookingId}`;
const availableKey   = () => `available:${EVENT_ID}`;

// ─── STEP 1: INITIALIZE SEATS ────────────────────────────────────────────────
async function initializeEvent() {
  console.log(`\n=== Initializing Event: ${EVENT_ID} ===`);
  const pipeline = redis.pipeline();

  for (let seat = 1; seat <= TOTAL_SEATS; seat++) {
    pipeline.set(seatStatusKey(seat), 'available');
  }
  // Store total available count
  pipeline.set(availableKey(), TOTAL_SEATS);
  await pipeline.exec();

  console.log(`${TOTAL_SEATS} seats initialized as available.`);
}

// ─── STEP 2: LOCK A SEAT (atomic SET NX PX) ──────────────────────────────────
/**
 * Returns true if lock acquired, false if seat already locked/booked.
 * Uses SET key value NX PX ttl  =>  atomic, no race condition.
 */
async function lockSeat(seat, userId) {
  const key   = seatLockKey(seat);
  const result = await redis.set(key, userId, 'NX', 'PX', LOCK_TTL_MS);
  return result === 'OK'; // OK = lock acquired, null = already locked
}

// ─── STEP 3: CONFIRM BOOKING (Lua atomic script) ─────────────────────────────
/**
 * Lua script ensures:
 *  1. The lock still belongs to this user (not expired and stolen)
 *  2. Seat status is still 'available'
 *  3. All three updates happen atomically:
 *     - Remove lock
 *     - Mark seat as 'booked'
 *     - Decrement available count
 */
const confirmBookingScript = `
  local lockKey    = KEYS[1]
  local statusKey  = KEYS[2]
  local availKey   = KEYS[3]
  local userId     = ARGV[1]

  -- Verify lock still belongs to this user
  local owner = redis.call('GET', lockKey)
  if owner ~= userId then
    return {err = 'LOCK_LOST: lock expired or taken by another user'}
  end

  -- Verify seat is still available
  local status = redis.call('GET', statusKey)
  if status ~= 'available' then
    return {err = 'SEAT_TAKEN: seat already booked'}
  end

  -- Atomically confirm booking
  redis.call('DEL', lockKey)
  redis.call('SET', statusKey, 'booked')
  redis.call('DECR', availKey)
  return 'OK'
`;

async function confirmBooking(seat, userId, bookingId) {
  try {
    const result = await redis.eval(
      confirmBookingScript,
      3,
      seatLockKey(seat),
      seatStatusKey(seat),
      availableKey(),
      userId
    );

    if (result === 'OK') {
      // Store booking details as a Redis Hash
      await redis.hset(bookingKey(bookingId), {
        bookingId,
        userId,
        seat,
        event:     EVENT_ID,
        status:    'confirmed',
        bookedAt:  new Date().toISOString()
      });
      // Set 24h expiry on booking record
      await redis.expire(bookingKey(bookingId), 86400);
      return { success: true, bookingId };
    }
  } catch (err) {
    return { success: false, reason: err.message };
  }
}

// ─── STEP 4: RELEASE LOCK (user cancelled selection) ─────────────────────────
async function releaseLock(seat, userId) {
  const lockKey = seatLockKey(seat);
  const owner   = await redis.get(lockKey);
  if (owner === userId) {
    await redis.del(lockKey);
    console.log(`  [RELEASED] Seat ${seat} lock released by ${userId}`);
    return true;
  }
  return false; // lock expired or not owned
}

// ─── STEP 5: CHECK SEAT STATUS ────────────────────────────────────────────────
async function getSeatStatus(seat) {
  const status  = await redis.get(seatStatusKey(seat));
  const lockedBy = await redis.get(seatLockKey(seat));
  return {
    seat,
    status: status || 'unknown',
    lockedBy: lockedBy || null
  };
}

// ─── STEP 6: GET ALL SEATS OVERVIEW ──────────────────────────────────────────
async function getEventOverview() {
  const available = await redis.get(availableKey());
  const seats = [];
  for (let s = 1; s <= TOTAL_SEATS; s++) {
    seats.push(await getSeatStatus(s));
  }
  return { event: EVENT_ID, availableCount: parseInt(available), seats };
}

// ─── STEP 7: CONCURRENT BOOKING SIMULATION ───────────────────────────────────
async function simulateConcurrentBookings() {
  console.log('\n=== Simulating Concurrent Bookings ===');

  // 15 users all try to book at the same time (only 10 seats available)
  const users = Array.from({ length: 15 }, (_, i) => ({
    userId:    `user-${i + 1}`,
    wantsSeat: (i % TOTAL_SEATS) + 1  // users spread across seats
  }));

  const bookAttempt = async ({ userId, wantsSeat }) => {
    const bookingId = `BKG-${Date.now()}-${userId}`;

    // 1. Try to lock the seat
    const locked = await lockSeat(wantsSeat, userId);
    if (!locked) {
      return { userId, seat: wantsSeat, result: 'FAILED', reason: 'Seat already locked' };
    }

    console.log(`  [LOCKED]  Seat ${wantsSeat} locked by ${userId}`);

    // 2. Simulate payment processing delay (100-300ms)
    await new Promise(r => setTimeout(r, Math.random() * 200 + 100));

    // 3. Confirm booking atomically
    const confirmation = await confirmBooking(wantsSeat, userId, bookingId);
    if (confirmation && confirmation.success) {
      console.log(`  [BOOKED]  Seat ${wantsSeat} confirmed for ${userId} (${bookingId})`);
      return { userId, seat: wantsSeat, result: 'SUCCESS', bookingId };
    } else {
      console.log(`  [FAILED]  Seat ${wantsSeat} booking failed for ${userId}: ${confirmation && confirmation.reason}`);
      return { userId, seat: wantsSeat, result: 'FAILED', reason: confirmation && confirmation.reason };
    }
  };

  // Fire all booking attempts concurrently
  const results = await Promise.all(users.map(bookAttempt));

  console.log('\n=== Booking Results ===');
  const successful = results.filter(r => r.result === 'SUCCESS');
  const failed     = results.filter(r => r.result === 'FAILED');
  console.log(`  Successful: ${successful.length}`);
  console.log(`  Failed    : ${failed.length}`);

  return results;
}

// ─── STEP 8: CANCEL A BOOKING ────────────────────────────────────────────────
async function cancelBooking(bookingId) {
  const booking = await redis.hgetall(bookingKey(bookingId));
  if (!booking || !booking.seat) {
    return { success: false, reason: 'Booking not found' };
  }

  // Free the seat
  await redis.set(seatStatusKey(booking.seat), 'available');
  await redis.incr(availableKey());
  await redis.hset(bookingKey(bookingId), 'status', 'cancelled');

  console.log(`  [CANCELLED] Booking ${bookingId} cancelled. Seat ${booking.seat} is now available.`);
  return { success: true };
}

// ─── MAIN ─────────────────────────────────────────────────────────────────────
(async () => {
  try {
    // Flush previous data for clean run
    await redis.flushall();

    await initializeEvent();
    await simulateConcurrentBookings();

    const overview = await getEventOverview();
    console.log('\n=== Final Seat Overview ===');
    console.log(`  Available: ${overview.availableCount} / ${TOTAL_SEATS}`);
    overview.seats.forEach(s => {
      const lock = s.lockedBy ? ` [locked by: ${s.lockedBy}]` : '';
      console.log(`  Seat ${s.seat}: ${s.status}${lock}`);
    });
  } finally {
    redis.disconnect();
  }
})();

File: ticket_booking.py
"""
Experiment-3: Concurrent Ticket Booking System with Seat Locking using Redis
Python Implementation

Install: pip install redis
Run    : python ticket_booking.py

Concepts demonstrated:
  - Atomic seat locking with SET NX PX
  - Lua scripts for atomic check-and-confirm
  - Concurrent booking simulation with ThreadPoolExecutor
  - Auto-expiring locks (TTL)
  - Booking stored as Redis Hash
"""

import redis
import time
import uuid
import random
from concurrent.futures import ThreadPoolExecutor, as_completed
from datetime import datetime

# ── CONFIG ────────────────────────────────────────────────────────────────────
LOCK_TTL_MS  = 10_000        # 10 seconds seat hold
EVENT_ID     = 'EVT-CONCERT-2026'
TOTAL_SEATS  = 10

# Redis connection
r = redis.Redis(host='localhost', port=6379, db=0, decode_responses=True)

# ── KEY HELPERS ───────────────────────────────────────────────────────────────
def seat_lock_key(seat):      return f"lock:{EVENT_ID}:seat:{seat}"
def seat_status_key(seat):    return f"status:{EVENT_ID}:seat:{seat}"
def booking_key(booking_id):  return f"booking:{booking_id}"
def available_key():          return f"available:{EVENT_ID}"

# ── STEP 1: INITIALIZE SEATS ─────────────────────────────────────────────────
def initialize_event():
    print(f"\n=== Initializing Event: {EVENT_ID} ===")
    pipe = r.pipeline()
    for seat in range(1, TOTAL_SEATS + 1):
        pipe.set(seat_status_key(seat), 'available')
    pipe.set(available_key(), TOTAL_SEATS)
    pipe.execute()
    print(f"{TOTAL_SEATS} seats initialized as available.")

# ── STEP 2: LOCK A SEAT (atomic SET NX PX) ───────────────────────────────────
def lock_seat(seat, user_id):
    """
    SET key value NX PX ttl
    Returns True if lock acquired, False if already locked.
    """
    result = r.set(seat_lock_key(seat), user_id, nx=True, px=LOCK_TTL_MS)
    return result is True

# ── STEP 3: CONFIRM BOOKING (Lua atomic script) ───────────────────────────────
CONFIRM_SCRIPT = """
  local lockKey   = KEYS[1]
  local statusKey = KEYS[2]
  local availKey  = KEYS[3]
  local userId    = ARGV[1]

  local owner = redis.call('GET', lockKey)
  if owner ~= userId then
    return redis.error_reply('LOCK_LOST')
  end

  local status = redis.call('GET', statusKey)
  if status ~= 'available' then
    return redis.error_reply('SEAT_TAKEN')
  end

  redis.call('DEL', lockKey)
  redis.call('SET', statusKey, 'booked')
  redis.call('DECR', availKey)
  return 'OK'
"""

_confirm_script = r.register_script(CONFIRM_SCRIPT)

def confirm_booking(seat, user_id, booking_id):
    try:
        result = _confirm_script(
            keys=[seat_lock_key(seat), seat_status_key(seat), available_key()],
            args=[user_id]
        )
        if result == 'OK':
            r.hset(booking_key(booking_id), mapping={
                'booking_id': booking_id,
                'user_id':    user_id,
                'seat':       seat,
                'event':      EVENT_ID,
                'status':     'confirmed',
                'booked_at':  datetime.utcnow().isoformat()
            })
            r.expire(booking_key(booking_id), 86400)  # 24h TTL
            return True, booking_id
        return False, 'UNKNOWN_ERROR'
    except redis.ResponseError as e:
        return False, str(e)

# ── STEP 4: RELEASE LOCK ─────────────────────────────────────────────────────
def release_lock(seat, user_id):
    lock_key = seat_lock_key(seat)
    owner = r.get(lock_key)
    if owner == user_id:
        r.delete(lock_key)
        print(f"  [RELEASED] Seat {seat} lock released by {user_id}")
        return True
    return False

# ── STEP 5: GET SEAT STATUS ───────────────────────────────────────────────────
def get_seat_status(seat):
    status    = r.get(seat_status_key(seat))
    locked_by = r.get(seat_lock_key(seat))
    return {'seat': seat, 'status': status or 'unknown', 'locked_by': locked_by}

# ── STEP 6: EVENT OVERVIEW ────────────────────────────────────────────────────
def get_event_overview():
    available = r.get(available_key())
    seats = [get_seat_status(s) for s in range(1, TOTAL_SEATS + 1)]
    return {'event': EVENT_ID, 'available_count': int(available or 0), 'seats': seats}

# ── STEP 7: SINGLE BOOKING ATTEMPT ───────────────────────────────────────────
def book_attempt(user_id, wants_seat):
    booking_id = f"BKG-{uuid.uuid4().hex[:8].upper()}-{user_id}"

    # 1. Try to lock
    locked = lock_seat(wants_seat, user_id)
    if not locked:
        return {'user_id': user_id, 'seat': wants_seat,
                'result': 'FAILED', 'reason': 'Seat already locked'}

    print(f"  [LOCKED]  Seat {wants_seat} locked by {user_id}")

    # 2. Simulate payment delay
    time.sleep(random.uniform(0.1, 0.3))

    # 3. Confirm atomically
    success, info = confirm_booking(wants_seat, user_id, booking_id)
    if success:
        print(f"  [BOOKED]  Seat {wants_seat} confirmed for {user_id} ({booking_id})")
        return {'user_id': user_id, 'seat': wants_seat,
                'result': 'SUCCESS', 'booking_id': booking_id}
    else:
        print(f"  [FAILED]  Seat {wants_seat} failed for {user_id}: {info}")
        return {'user_id': user_id, 'seat': wants_seat,
                'result': 'FAILED', 'reason': info}

# ── STEP 8: CONCURRENT SIMULATION ────────────────────────────────────────────
def simulate_concurrent_bookings():
    print('\n=== Simulating Concurrent Bookings ===')

    # 15 users competing for 10 seats
    users = [
        {'user_id': f'user-{i+1}', 'wants_seat': (i % TOTAL_SEATS) + 1}
        for i in range(15)
    ]

    results = []
    with ThreadPoolExecutor(max_workers=15) as executor:
        futures = [
            executor.submit(book_attempt, u['user_id'], u['wants_seat'])
            for u in users
        ]
        for f in as_completed(futures):
            results.append(f.result())

    successful = [r for r in results if r['result'] == 'SUCCESS']
    failed     = [r for r in results if r['result'] == 'FAILED']
    print(f"\n=== Booking Results ===")
    print(f"  Successful: {len(successful)}")
    print(f"  Failed    : {len(failed)}")
    return results

# ── STEP 9: CANCEL BOOKING ────────────────────────────────────────────────────
def cancel_booking(booking_id):
    booking = r.hgetall(booking_key(booking_id))
    if not booking or 'seat' not in booking:
        return False, 'Booking not found'
    seat = booking['seat']
    r.set(seat_status_key(seat), 'available')
    r.incr(available_key())
    r.hset(booking_key(booking_id), 'status', 'cancelled')
    print(f"  [CANCELLED] Booking {booking_id} cancelled. Seat {seat} is now available.")
    return True, 'OK'

# ── MAIN ──────────────────────────────────────────────────────────────────────
if __name__ == '__main__':
    r.flushall()  # clean slate

    initialize_event()
    simulate_concurrent_bookings()

    overview = get_event_overview()
    print(f"\n=== Final Seat Overview ===")
    print(f"  Available: {overview['available_count']} / {TOTAL_SEATS}")
    for s in overview['seats']:
        lock_info = f" [locked by: {s['locked_by']}]" if s['locked_by'] else ''
        print(f"  Seat {s['seat']}: {s['status']}{lock_info}")
